src/celeritas/user/ActionDiagnostic.cc

```diff
@@ -0,0 +1,250 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/ActionDiagnostic.cc
+//---------------------------------------------------------------------------//
+#include "ActionDiagnostic.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/data/CollectionAlgorithms.hh"
+#include "corecel/io/JsonPimpl.hh"
+#include "celeritas/global/ActionRegistry.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+#include "detail/ActionDiagnosticImpl.hh"
+
+#if CELERITAS_USE_JSON
+#    include <nlohmann/json.hpp>
+
+#    include "corecel/io/LabelIO.json.hh"
+#endif
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Helper function for accumulating data from host and device for all streams
+struct SumCounts
```

Comment:
Do you think it's reasonable to make this a more generic utility for integrating over memspace/streams, since we're now using a nearly identical operations between here and the SimpleCalo?

Reply:
Yes, I do (and this will be needed for the step diagnostic too).

Reply:
Hacked something in, though I'm sure there's a better way to do it...